ingest_by_manifest refactor

jcushman opened the pull request:
Some refactoring of ingest_by_manifest for performance.

- Ingest becomes a celery task. Instead of running a process pool, parallel work goes into subtasks handled by individual celery workers.
- More efficient redis use. We now have a "volumes" queue that just lists the name of each volume folder, a "volume:<volume_folder>" queue per volume folder, and an "errors" queue. Entries in a given inventory file are packed together so `read_inventory_file` only needs two redis calls per volume folder.
- Only validate the volmets against s3 file listing if we're fetching and updating volumexml.
- Optimize `ingest_volume` to run the minimum number of sql queries. Prefetch and cache all necessary data, only save if something changed.

I moved some things around in the process, so the code now pretty much reads from top to bottom, which made it easier for me to see any performance issues as I went along.

jcushman committed:
ingest_by_manifest refactor

ChefAndy commented:
Looks like redis isn't starting in the test env or something like that. All of the tests pass locally for me.

bensteinberg commented:
Do we need

     services:
      - redis-server

in `.travis.yml`?

jcushman committed:
Patch both redis dbs for tests

jcushman commented:
Sorry, fixed! We patch redis in a pytest fixture to (a) launch and (b) reset between each test, but I forgot to patch the second database connection that I added in this pull for use by the ingest script.

ChefAndy commented:
Nice. Ingest_by_manifest looks much cleaner.
